### PR TITLE
Release WDL, Feedback requested

### PR DESCRIPTION
**Pull request by ruchim**
It will not be merged into Cromwell, just assumed this was the easiest way to get it reviewed. It's simply a start at automating the release process, lots of feedback is welcome. 

Warning: Not been completely tested.

### CODE REVIEW

File: release.wdl [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+
+Array[Pair[String,Pair[String,String]]] releaseRepos = [("lenthall", ("0.20","0.21")),("wdl4s", ("0.8","0.9")),
+                                       ("wdltool", ("0.8","0.9")), ("cromwell", ("24","25"))]
+
+  scatter (repos in releaseRepos) {
+    call checkTests { input: baseDir = baseDir, repo=repos.left }
+  }
+
+  call basicRelease as lenthallRelease { input:dependdency = checkTests.result
```
Comment (cjllanwarne) [inline release.wdl]: typo in `input:dependdency`

File: release.wdl [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+   output{
+     String updatedVersion = releaseV
+   }
+}
+
+task wdltoolRelease {
+String baseDir
+String repo
+String releaseV
+String wdl4sV
```
Comment (cjllanwarne) [inline release.wdl]: Tabulation

File: release.wdl [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    #git push origin master
+    #git push --tags
+   }
+
+   output{
+     String updatedVersion = releaseV
+   }
+}
+
+task wdltoolRelease {
```
Comment (cjllanwarne) [inline release.wdl]: What's special about wdltool that it needs a separate block?

File: release.wdl [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+      if [[ -z $passingTests ]]
+      then
+        echo "Need to fix sbt tests for ${repo}" >&2
+        echo "Resuts of Sbt tests are: `cat results.txt`" >&2
+        exit 1
+      fi
+   }
+   output {
+     String results = read_string("results.txt")
```
Comment (cjllanwarne) [inline release.wdl]: We can use conditionals!
`Boolean success = ...`

**Comment by geoffjentry** [comment]
I'd double check what I say with @kshakir as he always corrects me but I'd vote for putting this in `src/bin/release`

File: release.wdl [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+        sed -i -e 's/'"$currentV"'/'"${releaseV}"'/' build.sbt
+        git add build.sbt
+        git commit -m "Update ${repo} version from $currentV to ${releaseV}"
+        if [ ${repo} -eq "wdl4s" ]; then #updating docs
+           sbt 'set scalacOptions in (Compile, doc) := List("-skip-packages", "better")' doc
+           git checkout gh-pages
+           mv target/scala-2.11/api ${releaseV}
+           git add ${releaseV}
+           git commit -m "API Docs"
+           #git push origin gh-pages
```
Comment (geoffjentry) [inline release.wdl]: I wonder if this is why that damn webpage always kept coming back ....

File: release.wdl [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+   git push origin develop
+
+  }
+  output {}
+}
+
+
+workflow release {
+
+String baseDir = "/Users/rmunshi"
```
Comment (geoffjentry) [inline release.wdl]: Eventually this shouldn't be `/Users/rmunshi` :)

File: release.wdl [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+     cd ${baseDir}/${repo}
+     git checkout develop
+     git pull
+
+     currentV=`grep "git.baseVersion " build.sbt | cut -d "=" -f 2 | tr -d \"\,`
+     echo "Version of ${repo} on the develop branch is $currentV"
+
+     if [ ${releaseV} -ne $version ]; then
+        echo "Need to update the ${repo} version on the develop branch to ${releaseV}" >&2
+        sed -i -e 's/'"$currentV"'/'"${releaseV}"'/' build.sbt
```
Comment (kshakir) [inline release.wdl]: Where environment is this script running? `sed` on macOS treats `-i` differently than say linux, and I didn't see a docker image specified.

http://stackoverflow.com/questions/5694228/sed-in-place-flag-that-works-both-on-mac-bsd-and-linux

Same goes for other `sed`s below.

File: release.wdl [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+String repo
+String releaseV
+String wdl4sV
+
+  command {
+    cd ${baseDir}/${repo}
+        git checkout develop
+        git pull
+
+        currentV=`grep "git.baseVersion " build.sbt | cut -d "=" -f2 | tr -d \"\,`
```
Comment (kshakir) [inline release.wdl]: Not sure if it's wdl friendly, but backticks are deprecated (here and other cases below). If it's compatible, use something like:

```bash
currentV=$(grep "git.baseVersion " build.sbt | cut -d "=" -f2 | tr -d \"\,)

File: release.wdl [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+
+
+task basicRelease {
+   Array[String] dependency
+   String repo
+   String releaseV
+   String baseDir
+
+   command {
+     cd ${baseDir}/${repo}
```
Comment (mcovarr) [inline release.wdl]: Perhaps start with:

```shell
set -e  # bomb out on the first non-zero return code
set -x  # print each command as it's executed
```

File: release.wdl [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+           git checkout develop
+           sed -i -e 's/'"$currentV"'/'"${releaseV}"'/' README.md
+           git add README.md
+           git commit -m "Update ${repo} API docs references"
+        fi
+        #git push origin develop
+     fi
+
+    #git checkout master
+    #git merge --squash origin/develop
```
Comment (mcovarr) [inline release.wdl]: I didn't think we were squash merging develop to master

File: release.wdl (hunk 5 above)
Comment (mcovarr) [inline release.wdl]: maybe dry run some of these git commands rather than comment them out?

File: release.wdl [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+
+  command {
+    cd ${baseDir}/${repo}
+    git checkout develop
+    git pull
+
+    currentV=`grep "cromwellVersion " project/Version.scala | cut -d "=" -f2 | tr -d \"`
+    echo "Version of ${repo} on the develop branch is $version"
+
+    if [ ${releaseV} -ne $version ];then
```
Comment (mcovarr) [inline release.wdl]: space after semicolon

File: release.wdl [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+task checkTests {
+  String baseDir
+  String repo
+
+   command {
+      cd ${baseDir}/${repo}
+      sbt test > results.txt
+      passingTests=`grep "All tests passed." results.txt`
```
Comment (mcovarr) [inline release.wdl]: This var should probably be more like `allTestsPassed`

File: release.wdl [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+
+      if [[ -z $passingTests ]]
+      then
+        echo "Need to fix sbt tests for ${repo}" >&2
+        echo "Resuts of Sbt tests are: `cat results.txt`" >&2
```
Comment (mcovarr) [inline release.wdl]: Resuts => Results
Sbt => sbt

File: release.wdl [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+  String baseDir
+  String repo
+  String upcomingV
+  String releasedV
+
+  command {
+   cd ${baseDir}/${repo}
+   git checkout develop
+
+   if [ ${repo} -eq "cromwell ]; then
```
Comment (mcovarr) [inline release.wdl]: unbalanced quote

File: release.wdl [hunk 15]
```diff
@@ -0,0 +1,238 @@
+#Thoughts--
+#Still need to manually check that the jars are being created in broad artifactory
+#Need to manually investigate if tests aren't being written
+#Need to manualy update the release portion of Cromwell
+#Need to make sure that the changelog.md file doesn't get weirdly altered when merging to master for Cromwell
```
Comment (geoffjentry) [inline release.wdl]: markdown doesn't auto-merge well it turns out. we ran into it when we did our release and @katevoss just ran into it with the wdl spec

File: release.wdl (hunk 5 above)
Comment (ruchim) [inline release.wdl]: I just didnt want to push any of my changes, makes me feel dirty

File: release.wdl [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+
+task basicRelease {
+   Array[String] dependency
+   String repo
+   String releaseV
+   String baseDir
+
+   command {
+     cd ${baseDir}/${repo}
+     git checkout develop
```
Comment (ruchim) [inline release.wdl]: The basic assumption I made is that entire WDL will be run locally on our laptops, and so for each task, you just need to provide the path for where the repo's are on your machine.

File: release.wdl (hunk 8 above)
Comment (Horneth) [inline release.wdl]: you might need to double the `$` to have WDL not complain

**Comment by ruchim** [comment]
Will reopen with some changes.

File: release.wdl [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+        if [ ${repo} -eq "wdl4s" ]; then #updating docs
+           sbt 'set scalacOptions in (Compile, doc) := List("-skip-packages", "better")' doc
+           git checkout gh-pages
+           mv target/scala-2.11/api ${releaseV}
+           git add ${releaseV}
+           git commit -m "API Docs"
+           #git push origin gh-pages
+
+           git checkout develop
+           sed -i -e 's/'"$currentV"'/'"${releaseV}"'/' README.md
```
Comment (ruchim) [inline release.wdl]: Update this command, it doesn't convert all necessary versions in the readme

**Pull request closed without merging**